mariadb: GRANT statement fails for multiple privileges - missing commas between privileges

### Describe the issue you are experiencing

After updating MariaDB addon to 3.0.0, the addon generates an invalid GRANT 
statement on every start. Privileges are separated by newlines instead of commas, 
causing ERROR 1064. HA recorder then gets "Access denied" on every restart.

### What type of installation are you running?

Home Assistant OS

### Which operating system are you running on?

Home Assistant Operating System

### Which app are you reporting an issue with?

MariaDB

### What is the version of the app?

3.0.0

### Steps to reproduce the issue

1. Update MariaDB addon to 3.0.0
2. Restart the addon
3. Check the addon log


### System Health information

## System Information

version | core-2026.4.3
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
container_arch | amd64
user | root
virtualenv | false
python_version | 3.14.2
os_name | Linux
os_version | 6.12.77-haos
arch | x86_64
timezone | Europe/Berlin
config_dir | /config

<details><summary>Home Assistant Community Store</summary>

GitHub API | ok
-- | --
GitHub Content | ok
GitHub Web | ok
HACS Data | ok
GitHub API Calls Remaining | 4982
Installed Version | 2.0.5
Stage | running
Available Repositories | 2965
Downloaded Repositories | 104

</details>

<details><summary>Home Assistant Cloud</summary>

logged_in | true
-- | --
subscription_expiration | 19. Dezember 2026 um 01:00
relayer_connected | true
relayer_region | eu-central-1
remote_enabled | true
remote_connected | true
alexa_enabled | true
google_enabled | false
cloud_ice_servers_enabled | true
remote_server | eu-central-1-36.ui.nabu.casa
certificate_status | ready
instance_id | b2c56db44f704cdfb9382dff6426dcff
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Home Assistant Supervisor</summary>

host_os | Home Assistant OS 17.2
-- | --
update_channel | stable
supervisor_version | supervisor-2026.04.0
agent_version | 1.8.1
docker_version | 29.3.1
disk_total | 80.0 GB
disk_used | 30.2 GB
nameservers | 10.10.5.253, 10.10.5.254
healthy | true
supported | true
host_connectivity | true
supervisor_connectivity | true
ntp_synchronized | true
virtualization | kvm
board | ova
supervisor_api | ok
version_api | ok
installed_addons | Matter Server (8.4.0), Cloudflared (7.0.5), Studio Code Server (6.0.1), Advanced SSH & Web Terminal (23.0.7), Samba share (12.6.1), NetworkManager Fix (1.1.0), Music Assistant (2.8.5), File editor (6.0.0), Glances (0.22.0), Zigbee2MQTT (2.9.2-1), AdGuard Home (6.1.3), AirSonos (5.1.1), OpenThread Border Router (2.16.7), Zigbee2MQTT (2.9.2-1), FTP (6.0.1), Speech-to-Phrase (1.4.3), go2rtc (1.9.14), Mosquitto broker (7.0.1), ESPHome Device Builder (2026.4.1), MariaDB (3.0.0), KNXD daemon (0.6.1), phpMyAdmin (0.13.0), eBUSd (26.1.8)

</details>

<details><summary>Dashboards</summary>

dashboards | 13
-- | --
resources | 64
views | 45
mode | storage

</details>

<details><summary>Network Configuration</summary>

adapters | lo (disabled), enp6s18 (disabled), enp6s19 (disabled), enp6s20 (enabled, default, auto), hassio (disabled), docker0 (disabled), veth4924134 (disabled), veth3696f4e (disabled), vethcf20c1a (disabled), veth64eb531 (disabled), veth6c1c997 (disabled), veth19db7d9 (disabled), veth3180c22 (disabled), veth98d3b11 (disabled), veth52ac3d1 (disabled), vethb5f3daa (disabled), vethd2459d5 (disabled), vethdf96942 (disabled), veth6e2fa7a (disabled), vethda70c60 (disabled), vethd0e4cf5 (disabled)
-- | --
ipv4_addresses | lo (127.0.0.1/8), enp6s18 (10.10.5.21/24), enp6s19 (10.10.3.21/24), enp6s20 (10.10.4.21/24), hassio (172.30.32.1/23), docker0 (172.30.232.1/23), veth4924134 (), veth3696f4e (), vethcf20c1a (), veth64eb531 (), veth6c1c997 (), veth19db7d9 (), veth3180c22 (), veth98d3b11 (), veth52ac3d1 (), vethb5f3daa (), vethd2459d5 (), vethdf96942 (), veth6e2fa7a (), vethda70c60 (), vethd0e4cf5 ()
ipv6_addresses | lo (::1/128), enp6s18 (fe80::d6ca:6ed1:9501:b81/64), enp6s19 (fe80::61b:ba2f:7e30:45a6/64), enp6s20 (fe80::30e:14b6:c9a1:6cbc/64), hassio (fd0c:ac1e:2100::1/48, fe80::60de:a7ff:fef0:c5cd/64), docker0 (fe80::1819:c8ff:fe2d:fe7/64), veth4924134 (fe80::48e1:91ff:fe8c:b415/64), veth3696f4e (fe80::200b:73ff:fe91:551c/64), vethcf20c1a (fe80::42d:29ff:fee4:8e44/64), veth64eb531 (fe80::18a1:73ff:fe3f:7fb3/64), veth6c1c997 (fe80::432:67ff:fef3:dfb1/64), veth19db7d9 (fe80::d860:7cff:fe75:7e2d/64), veth3180c22 (fe80::a8ec:c5ff:fe16:5ef1/64), veth98d3b11 (fe80::58be:c2ff:fef2:315/64), veth52ac3d1 (fe80::584a:4aff:fe5c:8269/64), vethb5f3daa (fe80::3451:f9ff:febe:98e2/64), vethd2459d5 (fe80::1446:adff:fec4:27a6/64), vethdf96942 (fe80::d4d0:95ff:fe50:1c86/64), veth6e2fa7a (fe80::30fe:e0ff:feff:bf1d/64), vethda70c60 (fe80::8c1a:65ff:fe25:324e/64), vethd0e4cf5 (fe80::3006:81ff:fe3f:5658/64)
announce_addresses | 10.10.4.21, fe80::30e:14b6:c9a1:6cbc

</details>

<details><summary>Recorder</summary>

oldest_recorder_run | 12. April 2026 um 22:15
-- | --
current_recorder_run | 20. April 2026 um 16:15
estimated_db_size | 5159.64 MiB
database_engine | mysql
database_version | 11.4.10

</details>


### Anything in the Supervisor logs that might be useful for us?

```txt
GRANT ALTER
CREATE
...
ERROR 1064 (42000) at line 1: You have an error in your SQL syntax; 
check the manual that corresponds to your MariaDB server version for 
the right syntax to use near 'CREATE
```

### Anything in the app logs that might be useful for us?

```txt
[15:57:47] INFO: Starting MariaDB with command line parameters: --innodb-buffer-pool-size=512M --innodb-log-file-size=128M --innodb-flush-log-at-trx-commit=2
2026-04-20 15:57:47 0 [Note] Starting MariaDB 11.4.10-MariaDB source revision 054a893f1645b77e52a329a7fc8cf614eebd1fad server_uid uT4FulGCZBOdXZVP10ICwriRhy8= as process 71
2026-04-20 15:57:47 0 [Note] InnoDB: Compressed tables use zlib 1.3.2
2026-04-20 15:57:47 0 [Note] InnoDB: Number of transaction pools: 1
2026-04-20 15:57:47 0 [Note] InnoDB: Using generic crc32 instructions
2026-04-20 15:57:47 0 [Note] InnoDB: Using Linux native AIO
2026-04-20 15:57:47 0 [Note] InnoDB: innodb_buffer_pool_size_max=512m, innodb_buffer_pool_size=512m
2026-04-20 15:57:47 0 [Note] InnoDB: Completed initialization of buffer pool
2026-04-20 15:57:47 0 [Note] InnoDB: Buffered log writes (block size=512 bytes)
2026-04-20 15:57:47 0 [Note] InnoDB: End of log at LSN=76092324896
2026-04-20 15:57:47 0 [Note] InnoDB: Opened 3 undo tablespaces
2026-04-20 15:57:47 0 [Note] InnoDB: 128 rollback segments in 3 undo tablespaces are active.
2026-04-20 15:57:47 0 [Note] InnoDB: Setting file './ibtmp1' size to 12.000MiB. Physically writing the file full; Please wait ...
2026-04-20 15:57:47 0 [Note] InnoDB: File './ibtmp1' size is now 12.000MiB.
2026-04-20 15:57:47 0 [Note] InnoDB: log sequence number 76092324896; transaction id 15268429
2026-04-20 15:57:47 0 [Note] Plugin 'FEEDBACK' is disabled.
2026-04-20 15:57:47 0 [Note] Plugin 'wsrep-provider' is disabled.
2026-04-20 15:57:47 0 [Note] InnoDB: Loading buffer pool(s) from /data/databases/ib_buffer_pool
2026-04-20 15:57:47 0 [Note] InnoDB: Buffer pool(s) load completed at 260420 15:57:47
2026-04-20 15:57:48 0 [Note] Server socket created on IP: '0.0.0.0', port: '3306'.
2026-04-20 15:57:48 0 [Note] Server socket created on IP: '::', port: '3306'.
2026-04-20 15:57:48 0 [Note] mariadbd: Event Scheduler: Loaded 0 events
2026-04-20 15:57:48 0 [Note] mariadbd: ready for connections.
Version: '11.4.10-MariaDB'  socket: '/run/mysqld/mysqld.sock'  port: 3306  Alpine Linux
[15:57:49] INFO: MariaDB started
[15:57:49] INFO: Check data integrity and fix corruptions
mysql.column_stats                                 OK
mysql.columns_priv                                 OK
mysql.db                                           OK
mysql.event                                        OK
mysql.func                                         OK
mysql.global_priv                                  OK
mysql.gtid_slave_pos                               OK
mysql.help_category                                OK
mysql.help_keyword                                 OK
mysql.help_relation                                OK
mysql.help_topic                                   OK
mysql.index_stats                                  OK
mysql.innodb_index_stats                           OK
mysql.innodb_table_stats                           OK
mysql.plugin                                       OK
mysql.proc                                         OK
mysql.procs_priv                                   OK
mysql.proxies_priv                                 OK
mysql.roles_mapping                                OK
mysql.servers                                      OK
mysql.table_stats                                  OK
mysql.tables_priv                                  OK
mysql.time_zone                                    OK
mysql.time_zone_leap_second                        OK
mysql.time_zone_name                               OK
mysql.time_zone_transition                         OK
mysql.time_zone_transition_type                    OK
mysql.transaction_registry                         OK
homeassistant.event_data                           OK
homeassistant.event_types                          OK
homeassistant.events                               OK
homeassistant.migration_changes                    OK
homeassistant.recorder_runs                        OK
homeassistant.schema_changes                       OK
homeassistant.stat_mapping                         OK
homeassistant.state_attributes                     OK
homeassistant.states                               OK
homeassistant.states_meta                          OK
homeassistant.statistics                           OK
homeassistant.statistics_meta                      OK
homeassistant.statistics_runs                      OK
homeassistant.statistics_short_term                OK
phpmyadmin.pma__bookmark                           OK
phpmyadmin.pma__central_columns                    OK
phpmyadmin.pma__column_info                        OK
phpmyadmin.pma__designer_settings                  OK
phpmyadmin.pma__export_templates                   OK
phpmyadmin.pma__favorite                           OK
phpmyadmin.pma__history                            OK
phpmyadmin.pma__navigationhiding                   OK
phpmyadmin.pma__pdf_pages                          OK
phpmyadmin.pma__recent                             OK
phpmyadmin.pma__relation                           OK
phpmyadmin.pma__savedsearches                      OK
phpmyadmin.pma__table_coords                       OK
phpmyadmin.pma__table_info                         OK
phpmyadmin.pma__table_uiprefs                      OK
phpmyadmin.pma__tracking                           OK
phpmyadmin.pma__userconfig                         OK
phpmyadmin.pma__usergroups                         OK
phpmyadmin.pma__users                              OK
sys.sys_config                                     OK
[15:57:49] INFO: Ensuring internal database upgrades are performed
[15:57:49] INFO: Ensure databases exists
[15:57:49] INFO: Create database homeassistant
[15:57:49] INFO: Ensure users exists and are updated
[15:57:49] INFO: Update user homeassistant
[15:57:49] INFO: Init/Update rights
[15:57:50] INFO: Granting ALTER
CREATE
CREATE ROUTINE
CREATE TEMPORARY TABLES
CREATE VIEW
DELETE
DELETE HISTORY
DROP
EVENT
INDEX
INSERT
LOCK TABLES
SELECT
SHOW VIEW
TRIGGER
UPDATE to homeassistant on homeassistant
--------------
GRANT ALTER
CREATE
CREATE ROUTINE
CREATE TEMPORARY TABLES
CREATE VIEW
DELETE
DELETE HISTORY
DROP
EVENT
INDEX
INSERT
LOCK TABLES
SELECT
SHOW VIEW
TRIGGER
UPDATE ON homeassistant.* TO 'homeassistant'@'%'
--------------
ERROR 1064 (42000) at line 1: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'CREATE
CREATE ROUTINE
CREATE TEMPORARY TABLES
CREATE VIEW
DELETE
DELETE HISTO...' at line 2
[15:57:50] INFO: Sending service information to Home Assistant
s6-rc: info: service mariadb-post successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
2026-04-20 15:59:19 18 [Warning] Access denied for user 'root'@'172.30.32.1' (using password: YES)
2026-04-20 15:59:29 19 [Warning] Access denied for user 'root'@'172.30.32.1' (using password: YES)
2026-04-20 16:04:26 20 [Warning] Access denied for user 'homeassistant'@'%' to database 'homeassistant'
2026-04-20 16:04:57 21 [Warning] Access denied for user 'root'@'172.30.32.1' (using password: YES)
```

### Additional information

Workaround: 
docker exec addon_core_mariadb mariadb -u root -e "GRANT ALTER, CREATE, 
CREATE ROUTINE, CREATE TEMPORARY TABLES, CREATE VIEW, DELETE, DELETE HISTORY, 
DROP, EVENT, INDEX, INSERT, LOCK TABLES, SELECT, SHOW VIEW, TRIGGER, UPDATE 
ON homeassistant.* TO 'homeassistant'@'%'; FLUSH PRIVILEGES;"

I see where's the problem, but I disagree it's a 3.0.0 regression - it seems it was a pre-existing issue with 2.7.2 as well if the `privileges` contained more than one permission with a config like this:
```yaml
databases:
  - homeassistant
logins:
  - password: homeassistant
    username: homeassistant
  - username: username
    password: password
rights:
  - database: homeassistant
    username: homeassistant
  - username: username
    privileges:
      - ALTER
      - CREATE
    database: homeassistant
```

This is a 2.7.2 boot log:
```
[18:55:10] INFO: Starting MariaDB
2026-04-20 18:55:10 0 [Note] Starting MariaDB 10.11.6-MariaDB source revision fecd78b83785d5ae96f2c6ff340375be803cd299 as process 67
2026-04-20 18:55:10 0 [Note] InnoDB: Compressed tables use zlib 1.3.1
2026-04-20 18:55:10 0 [Note] InnoDB: Number of transaction pools: 1
2026-04-20 18:55:10 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
2026-04-20 18:55:10 0 [Note] InnoDB: Using Linux native AIO
2026-04-20 18:55:10 0 [Note] InnoDB: Initializing buffer pool, total size = 128.000MiB, chunk size = 2.000MiB
2026-04-20 18:55:10 0 [Note] InnoDB: Completed initialization of buffer pool
2026-04-20 18:55:10 0 [Note] InnoDB: Resetting space id's in the doublewrite buffer
2026-04-20 18:55:10 0 [Note] InnoDB: Buffered log writes (block size=512 bytes)
2026-04-20 18:55:10 0 [Note] InnoDB: End of log at LSN=61565
2026-04-20 18:55:10 0 [ERROR] InnoDB: The change buffer is corrupted or has been removed on upgrade to MariaDB 11.0 or later
2026-04-20 18:55:10 0 [Note] InnoDB: 128 rollback segments are active.
2026-04-20 18:55:10 0 [Note] InnoDB: Setting file './ibtmp1' size to 12.000MiB. Physically writing the file full; Please wait ...
2026-04-20 18:55:10 0 [Note] InnoDB: File './ibtmp1' size is now 12.000MiB.
2026-04-20 18:55:10 0 [Note] InnoDB: log sequence number 61565; transaction id 14
2026-04-20 18:55:10 0 [Note] Plugin 'FEEDBACK' is disabled.
2026-04-20 18:55:10 0 [Note] InnoDB: Loading buffer pool(s) from /data/databases/ib_buffer_pool
2026-04-20 18:55:10 0 [Note] Server socket created on IP: '0.0.0.0'.
2026-04-20 18:55:10 0 [Note] Server socket created on IP: '::'.
2026-04-20 18:55:10 0 [Note] InnoDB: Buffer pool(s) load completed at 260420 18:55:10
2026-04-20 18:55:10 0 [Note] mysqld: ready for connections.
Version: '10.11.6-MariaDB'  socket: '/run/mysqld/mysqld.sock'  port: 3306  Alpine Linux
[18:55:11] INFO: MariaDB started
[18:55:11] INFO: Check data integrity and fix corruptions
mysql.column_stats                                 OK
mysql.columns_priv                                 OK
mysql.db                                           OK
mysql.event                                        OK
mysql.func                                         OK
mysql.global_priv                                  OK
mysql.gtid_slave_pos                               OK
mysql.help_category                                OK
mysql.help_keyword                                 OK
mysql.help_relation                                OK
mysql.help_topic                                   OK
mysql.index_stats                                  OK
mysql.innodb_index_stats                           OK
mysql.innodb_table_stats                           OK
mysql.plugin                                       OK
mysql.proc                                         OK
mysql.procs_priv                                   OK
mysql.proxies_priv                                 OK
mysql.roles_mapping                                OK
mysql.servers                                      OK
mysql.table_stats                                  OK
mysql.tables_priv                                  OK
mysql.time_zone                                    OK
mysql.time_zone_leap_second                        OK
mysql.time_zone_name                               OK
mysql.time_zone_transition                         OK
mysql.time_zone_transition_type                    OK
mysql.transaction_registry                         OK
sys.sys_config                                     OK
[18:55:11] INFO: Ensuring internal database upgrades are performed
[18:55:11] INFO: Ensure databases exists
[18:55:11] INFO: Create database homeassistant
[18:55:11] INFO: Ensure users exists and are updated
[18:55:11] INFO: Update user homeassistant
[18:55:11] INFO: Update user username
[18:55:11] INFO: Init/Update rights
[18:55:11] INFO: Granting all privileges to homeassistant on homeassistant
[18:55:11] INFO: Granting ALTER
CREATE to username on homeassistant
ERROR 1064 (42000) at line 1: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'CREATE ON homeassistant.* TO 'username'@'%'' at line 2
[18:55:11] INFO: Sending service information to Home Assistant
s6-rc: info: service mariadb-post successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
```

Anyway, desires a fix :)

--- a/mariadb/CHANGELOG.md
+++ b/mariadb/CHANGELOG.md
@@ -1,5 +1,9 @@
 # Changelog
 
+## 3.0.1
+
+- Fix GRANT statement generation for multiple `privileges`
+
 ## 3.0.0
 
 **Note:** This update upgrades MariaDB from 10.11 to 11.4. The database

--- a/mariadb/config.yaml
+++ b/mariadb/config.yaml
@@ -1,5 +1,5 @@
 ---
-version: 3.0.0
+version: 3.0.1
 breaking_versions: [3.0.0]
 slug: mariadb
 name: MariaDB

--- a/mariadb/rootfs/etc/s6-overlay/s6-rc.d/mariadb-post/run
+++ b/mariadb/rootfs/etc/s6-overlay/s6-rc.d/mariadb-post/run
@@ -72,7 +72,7 @@ for right in $(bashio::config "rights|keys"); do
     DATABASE=$(bashio::config "rights[${right}].database")
 
     if bashio::config.exists "rights[${right}].privileges"; then
-        PRIVILEGES=$(bashio::config "rights[${right}].privileges")
+        PRIVILEGES=$(bashio::config "rights[${right}].privileges | (if type == \"array\" then join(\", \") else . end)")
         bashio::log.info "Granting ${PRIVILEGES} to ${USERNAME} on ${DATABASE}"
         mariadb -e "REVOKE ALL PRIVILEGES ON ${DATABASE}.* FROM '${USERNAME}'@'%';" || true
         mariadb -e "GRANT ${PRIVILEGES} ON ${DATABASE}.* TO '${USERNAME}'@'%';" || true

